### ISSUE TITLE
extracting input features that contribute to class activation keras r

### ISSUE THREAD

**Issue by amjass12** [rstudio/keras#756]
Hello all, 

Would it be possible to push me in the right direction for a little issue I am having. I am working with tabular data in order to classify some genomic data such as sex, age etc. There are 5000 features that contribute to and define these features. I have run a multi-label sequential model as follows:

one-hot encode label data: 6 classes of data:

```
model2 <- keras_model_sequential()


model2 %>%
   layer_dense(units=150, activation = 'relu', input_shape = 5000) %>%  
   layer_dense(units = 64, activation = "relu") %>%
   layer_dense(units=ncol(trainingtarget1), activation = "sigmoid")


model2 %>% keras::compile(loss='binary_crossentropy',
                          optimizer='adam',
                          metrics='accuracy')


history2 <- model2%>% 
   fit(as.matrix(training1), 
       as.matrix(trainingtarget1), 
       epoch=200,
       batch=32,
       validation_split = 0.15,
       callbacks = list(early_stop, print_dot_callback))

model2%>%
   keras::evaluate(training1,as.matrix(trainingtarget1))
model2%>%
   keras::evaluate(test1,as.matrix(testtarget1))
```

This model performs well and confusion matrix looks good for the different classes! However, now I want to be able to understand which input features (of the 5000) are contributing to different activations:

I have separated the layers in order to see how they perform 

```
layer_name<-"dense_1"
intermediate_layer_model <- keras_model(inputs = model2$input, outputs = get_layer(model2, layer_name)$output)
                                        
intermediate_output <- predict(intermediate_layer_model, training1)
```

I can plot PCA's and heatmaps that show nicely where neurons are lighting up etc, however, what I cannot extract are the input features at each unit that are causing them light up! In other words, I cannot seem to circle back to the original training data to see which of the 5000 features are being considered as important at the different units/for different classifications!

Is this the role of the get_input_at function? 

Any advice would be much appreciated!
Many thanks!

**Comment by skeydan** [comment]
You could retrieve complete tensors, like so:

```
l <- intermediate_layer_model %>% get_layer(layer_name)

# there is just one input node, so index = 0
# shape=(?, 150)
l$get_input_at(0L) 
```

**Comment by amjass12** [comment]
Thank you!  I have done this but am still slightly confused...  I get the following output:

Tensor("dense/Relu:0", shape=(?, 150), dtype=float32)

How do I now retrieve the actual tensor and will it give me information about the feature and their weights at respective neurons? I am really sorry, I am so confused about this particular problem, but this is actually the information that is most important! I am pasting a heatmap of this hidden layer. I would like to see which features are causing the units to fire up (red activations) to see how these features are shared and what network is considering as important in beginning to classify the data.

This is the one thing I just simply cant figure out as so much of the code is focused on hidden layers as oppose to individual elements (at leats that is what I cant figure out!)

Thanks!


<img width="773" alt="Screenshot 2019-04-25 at 08 44 55" src="https://user-images.githubusercontent.com/33659783/56718361-7fb56500-6736-11e9-9cd3-aa4df8473f14.png">

**Comment by skeydan** [comment]
I think what you might be interested in are the layer's `weights`.

Assuming for a moment you were using a linear activation function, all that would happen would be a matrix multiplication between inputs and weights, resulting in the activations you're already tracking...
Put differently, the weights provide an indication what connections the network deems important.

**Comment by amjass12** [comment]
Yes!! I believe this is it!! Attaching a screenshot! But as the model gets rid of row and column names, is there a way of defining which specific features from the training data it deems important, the first layer is easy as it is all features, but more difficult downstream...

also, does dictating a hidden layer of 150 nodes mean that only 150 features will be considered in the hidden layer from all 5042 from it? or are  different features contained within those 150 nodes? The attached screenshot would suggest there are only 150 features in each node

Thank you for helping me with this! i do appreciate it!


<img width="1018" alt="Screenshot 2019-04-25 at 09 46 14" src="https://user-images.githubusercontent.com/33659783/56722745-99a77580-673f-11e9-8c99-255a1779e41f.png">

**Comment by skeydan** [comment]
>  is there a way of defining which specific features from the training data it deems important, the first layer is easy as it is all features, but more difficult downstream...

that somehow is the whole point of neural networks, that things get "mixed up" ... otherwise you might use a straightforward (logistic) regression...

> also, does dictating a hidden layer of 150 nodes mean that only 150 features will be considered in the hidden layer from all 5042 from it?

If we assume you stopped at that first layer, and used a linear activation, the network would do the same thing as regression: It combines the inputs linearly. So if that layer had 1 unit, you'd have linear regression, and that too does not choose one input feature, but combines all of them.
Really in that case, the output is nothing but matrix multiplication between the layer weights and the input (which means all input features are considered, just as in regression).

**Comment by amjass12** [comment]
Thank you. Yes that makes sense! But then this deepens the mystery for me! If the heatmap above suggest neurons being lit up at some node for some samples etc... I cannot actually access their weights. The only weights I can access are from the first layer (5042 passed on to 150)... But how would I identify the features that are being passed on to the next hidden layer (64) before the final output?

**Comment by skeydan** [comment]
Honestly I don't think that makes a lot of sense... Technically, yes, you can access weights from all intermediate layers. But I don't see how that helps.

> But how would I identify the features that are being passed on to the next hidden layer (64) before the final output?

There's just no _features_ any more, it's all gotten mangled up...

 If it's about interpretability etc. I think you would need to look into other methods specifically designed for that. Perhaps this book

https://christophm.github.io/interpretable-ml-book/

can help you. But it's a vast field...

**Comment by amjass12** [comment]
So the reason for this would be to probe shared weights: so in the heatmap above the there are nodes that light up in samples that have common features (common categorical data such as gender). Here is where I would want to see, what is it that lights them up and what is shared across samples...and which features the networks believes are important.  although mangled up, its the features that one would be interested in in understanding whats driving certain characteristics. 

It would be like accessing the final layer and seeing which features drive those classifications.. but by understanding which are common between samples, that would be useful. There must be a way of of . asking what is causing classifications and relationships. Even without interpretability methods, probing the hidden layers and even the final layer should provide insight?

**Comment by skeydan** [comment]
unfortunately I don't really think so, but you might try asking in other places, such as Quora or StackOverflow

**Comment by amjass12** [comment]
Thank you for all your help today!

**Comment by skeydan** [comment]
You're welcome!
